File: leetcode/algorithms/easy/min-stack.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Min Stack"
*/


/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.min = null;
    this.stack = [];
    // this.topOfStack=-1;
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    this.stack.push(x);
    // this.topOfStack++;
    if(this.min===null){
        this.min=x;
    } else {
        this.min = Math.min(this.min,x);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if(this.stack.length===0){
        return;
    }
    this.stack.pop();
    // this.topOfStack--;
    if(this.stack.length===0){
        this.min=null;
    } else {
        this.min = Math.min(...this.stack);
    }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    if(this.stack.length===0){
        return null;
    } else {
        return this.stack[this.stack.length-1];
    }
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.min;
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: leetcode/algorithms/stack/min-stack.py
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        
        self.stack = [];
        self.min = None;
        

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        if self.min==None:
            self.min = val
            
        if val<=self.min:
            self.stack.append(self.min);
            self.min=val;
        
        self.stack.append(val);
            
        

    def pop(self):
        """
        :rtype: None
        """
        
        poppedVal = self.stack.pop();
        if poppedVal==self.min:
            self.min = self.stack.pop();
            
        if len(self.stack)==0:
            self.min=None
        
    def top(self):
        """
        :rtype: int
        """
        return self.stack[len(self.stack)-1];
        

    def getMin(self):
        """
        :rtype: int
        """
        return self.min;
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()